### DISCUSSION TITLE
Help with non-build-time variables

### DISCUSSION THREAD

**Discussion by merisbahti** [vercel/next.js#19605] [Help]
Suppose that I have several server deployments, ci, staging, and prod. And they all have different secrets that are to be used in the client. But I have only 1 build. Is it possible to have different sets of .env variables for each deployment, but only 1 built artifact?

I've tried making a `.env` file with the appropriate env variables set, but in the browser bundle that's served to the client, only build time variables are avaliable.

The new variables from the `.env` file are avaliable in my custom `server.js` when I log `process.env.VARIABLENAME`

**Comment by lxsmnsyc** [comment]
Yes, only build time variables are available, as the environment variables are only read once during build time (since the web has no concept of "process.env"). If you have different environments, I think something like `.env.development` and `.env.production` should work but if it doesn't cover all of the cases, you might want to provide a custom `next.config.js` and provide the necessary variables to `config.env`:

```js
module.exports = {
  env: {
    MY_VARIABLE: process.env.NODE_ENV === 'production' ? process.env.MY_VARIABLE_PRODUCTION : process.env.MY_VARIABLE_DEVELOPMENT, 
  },
};

**Comment by michrome** [comment]
@merisbahti you can [expose environment variables to the browser](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser) using the naming convention `NEXT_PUBLIC_XXX`.

I have [a good example you can copy](https://github.com/vercel/next.js/discussions/19576) as I asked this question earlier this week.
